core/multigrid/amgx_pgm.cpp

```diff
@@ -141,27 +142,34 @@ void AmgxPgm<ValueType, IndexType>::generate()
 
     gko::dim<2>::dimension_type coarse_dim = num_agg;
     auto fine_dim = system_matrix_->get_size()[0];
-    // TODO: prolong_op can be done with lightway format
-    auto prolong_op = share(
+    // prolong_row_gather is the lightway implementation for prolongation
+    // TODO: However, we still create the csr to process coarse/restrict matrix
+    // generation. It may be changed when we have the direct triple product from
+    // agg index.
+    auto prolong_row_gather = share(matrix::RowGatherer<IndexType>::create(
+        exec, gko::dim<2>{fine_dim, coarse_dim}));
+    exec->copy_from(exec.get(), agg_.get_num_elems(), agg_.get_const_data(),
+                    prolong_row_gather->get_row_gather_index());
+    auto prolong_csr = share(
         matrix_type::create(exec, gko::dim<2>{fine_dim, coarse_dim}, fine_dim));
     exec->copy_from(exec.get(), agg_.get_num_elems(), agg_.get_const_data(),
-                    prolong_op->get_col_idxs());
-    exec->run(amgx_pgm::make_fill_seq_array(prolong_op->get_row_ptrs(),
+                    prolong_csr->get_col_idxs());
+    exec->run(amgx_pgm::make_fill_seq_array(prolong_csr->get_row_ptrs(),
                                             fine_dim + 1));
-    exec->run(amgx_pgm::make_fill_array(prolong_op->get_values(), fine_dim,
+    exec->run(amgx_pgm::make_fill_array(prolong_csr->get_values(), fine_dim,
                                         one<ValueType>()));
     // TODO: implement the restrict_op from aggregation.
```

Comment:
I guess because of this transpose operation, this generation is slower than AMGX ?

Reply:
yes, and it uses two csr spgemm to generate coarse